### ISSUE TITLE
Query containing hasOne association pointing to a class having the StiParent behavior triggers error if no associated data is found

### ISSUE THREAD

**Issue by k-risc** [robotusers/cakephp-table-inheritance#10]
The `StiParentBehavior`, in its `beforeFind()` method, registers a formatter with `formatResults()`, in which it does its magic.

If I run a query on some other main table, having a `hasOne` association defined to the table with the `StiParentBehavior`, and I contain that association, then the following error occurs if one of the results does not have such an `hasOne` related record:
`Argument 1 passed to Robotusers\TableInheritance\Model\Behavior\StiParentBehavior::Robotusers\TableInheritance\Model\Behavior\{closure}() must implement interface Cake\Datasource\EntityInterface, null given`

The reason:
* The default strategy of hasOne associations is `join`
* Thus, the sql statement for the main query gets additional code added for a JOIN to the associated table, so their fields are directly loaded with the main query
* In the `Association` class, the method `_formatAssociationResults()` is executed, amongst others, with the following code:
```
$property = $options['propertyPath'];
$propertyPath = explode('.', $property);
$query->formatResults(function ($results, $query) use ($formatters, $property, $propertyPath) {
    $extracted = [];
    foreach ($results as $result) {
        foreach ($propertyPath as $propertyPathItem) {
            if (!isset($result[$propertyPathItem])) {
                $result = null;
                break;
            }
            $result = $result[$propertyPathItem];
        }
        $extracted[] = $result;
    }
    $extracted = new Collection($extracted);
    foreach ($formatters as $callable) {
        $extracted = new ResultSetDecorator($callable($extracted, $query));
    }
    ...
```
* If there is no hasOne related record, then `$result[$propertyPathItem]` is null. This leads to `$result` being set to `null`, and that being added to `$extracted`. For a hasOne association, as it can be max. one result, `$extracted`will then be:
```
[
    0 => null,
]
``` 
* This is then sent to each formatter. One of which being the `formatResults()` defined in the `beforeFind()` of  `StiParentBehavior`.
* It does a `$results->map(function (EntityInterface $row) {...})`. `$results`only has one element, and that has the value `null`, so consequently the error occurs, because the function within  `$results->map(...)` expects the parameter `$row` to implement the `EntityInterface´.

If I change the `strategy` of the `hasOne` assocation to `select`. Then this works. If the associated data is retrieved with a separate query, then `_formatAssociationResults()` is not executed, and the error is not triggered.

I am wondering, is that a conceptual issue, an issue on the way CakePHP 4 does with hasOne associated data if it is JOINed to the main query? Or would it be feasible to amend the `StiParentBehavior`? I am not sure if that could cause any other problems somewhere else, or whether it was done for a particular reason to force $row to implement the EntityInterface and not be null.

**Comment by robertpustulka** [comment]
Thanks for the report. You can provide a PR with a patch

**Comment by k-risc** [comment]
I believe that my set up was wrong. I had a parent table class with the StiParent behavior loaded, and I extended a "child" table class from this parent table. After gaining some more experience, I know that the child tables must only get the Sti behavior attached, so they should not be extended from the parent table class.

I will close this.

The only issue I found which is similar to the above is that if you use the "list" finder on a table with the StiParent behavior, you will get the same error as mentioned above. formatResults() and map() will be called as well, this time with the $row being whatever the "value" part of the list results is. Most of the times a string, but it could yield an integer, a FrozenDate or something else, depending on the further settings when calling the list finder. This can be circumvented be disabling hydration before calling the finder. I would assume that CakePHP does not do it automatically for the list finder because it still makes sense to first hydrate the results into ORM classes instances, as the value field chosen for the list results van be a virtual field, and in that case we need the getter function in the ORM class in order to retrieve the final value. Hence disabling hydration when using the list finder is not always an option.

But I think the underlying problem is that findList also uses formatResults(), and having a query which causes multiple formatResults() to be executed is tricky, as each of them might change the $results in such a fundamental way that they are not compatible any more with the next call to formatResults().
